text/0012-milestone-merkle-validation/0012-milestone-merkle-validation.md

```diff
@@ -0,0 +1,124 @@
+
+- Perform tip selection to choose a branch and a trunk for the new milestone.
+- Determine the topological order according to [RFC-0005](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0005-white-flag/0005-white-flag.md) of the referenced bundles that are not yet confirmed by a previous milestone.
+- Construct the list B<sup>tri</sup> consisting of the tail transaction hashes of all the not-ignored state-mutating bundles in that particular order. A bundle is state-mutating, if it actually modifies the ledger state by moving funds from one address to another. That is, any bundle having an address with a total output value larger than that address' total input value, is a state-mutating bundle.
+- Convert each element of B<sup>tri</sup> to binary by applying the `t5b1` encoding. (This is exactly the conversion used for binary I/O of ternary data in the current protocol.) This leads to the ordered list B containing 49-byte strings.
+- Compute the 64-byte Merkle tree hash H = MTH(B).
+- Convert H to ternary by applying the encoding described in [RFC-0015](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0015-binary-to-ternary-encoding/0015-binary-to-ternary-encoding.md). This leads to H<sup>tri</sup> with a length of 384 trits.
+- Prepare the milestone bundle as usual. Its head transaction contains the information required to verify the Coordinator's signature in its `signatureMessageFragment` field. This information has a length of d·81 trytes, where d is the depth of the Coordinator's Merkle tree. 
+- Append H<sup>tri</sup> to the `signatureMessageFragment` field. For any depth d < 26 the field provides sufficient space.
+- Sign the head transaction and add its fragmented signature to the milestone bundle's zero value transactions.
+
+## Milestone validation
+
+- Verify the signature of the milestone m.
+- Construct the ordered list B<sup>tri</sup> of the tail transaction hashes of the not-ignored state-mutating bundles m confirms.
+- Encode the hashes B<sup>tri</sup> into their binary representation B and compute H = MTH(B).
+- Extract the first 384 trits after the Coordinator's Merkle tree information from the `signatureMessageFragment` field of the head transaction and verify that this matches the ternary encoded H.
+
+## Proof of inclusion
+
+- Identify the confirming milestone m of the state-mutating input bundle b.
+- Determine the ordered list of the not-ignored bundles m confirms.
+- Compute the Merkle audit path of b with respect to the Merkle tree for this ordered list.
+- Provide the audit path as well as m as proof of inclusion for b.
+
+## Cryptographic components
+
+### Merkle hash trees
+
+This RFC uses a binary Merkle hash tree for efficient auditing. In general, any cryptographic hashing algorithm can be used for this. However, we propose to use [BLAKE2b-512](https://tools.ietf.org/html/rfc7693), as it provides a faster and more secure alternative to the widely used SHA-256/SHA-512. 
```

Comment:
Judging from other future proposals, it seems like BLAKE2b-256 is going to be used for things like transaction hashes in the future. With this in mind, should we also change here to use the 256-bit hash instead of 512-bit?
#### Con
- Current transaction hashes are effectively 384-bit. Using a smaller hash for the Merkle Tree would reduce the entropy even compared to a single transaction.
- Future transactions hash formats are not yet finalized.
- BLAKE2b-256 is exactly the same hash function as BLAKE2b-512 only with the first bits considered.
#### Pro
- Potentially reduce the number of used hash functions/versions in future.

Reply:
Keep BLAKE2b-512 for now